### ISSUE TITLE
getIdToken throws "auth/user-token-expired" if refresh token is revoked

### ISSUE THREAD

**Issue by jaunt** [firebase/firebase-js-sdk#2886]
* Operating System version: iOS 13.3.1
  * Browser version: Safari Built-In
  * Firebase SDK version: 7.3.0
  * Firebase Product: auth

In order to communicate to my server from a client I use:

```
        firebase
            .auth()
            .currentUser.getIdToken(true)
            .then(idToken => {
              callServer(idToken);
            })
            .catch(error => {
              if (error.code === "auth/user-token-expired") {
                logError(error, "user-token-expired");
```

Occasionally, I catch the error auth/user-token-expired above.

Yet the documentation says that this function _"Returns the current token if it has not expired. Otherwise, this will refresh the token and return a new one."_

From that description, I'm surprised it ever throws that error?  How should I handle the situation?  Recursively call itself?  Seems like a bad idea.

It's hard to reproduce unfortunately, even though I do get multiple reports from customers.  It seems to happen mostly or possibly only iOS safari.  Perhaps it's when the tab is returned to after being dormant, but I'm not sure.

Any help would be greatly appreciated.

**Comment by hsubox76** [comment]
From looking at the code (https://firebase.google.com/docs/auth/admin/manage-sessions) I think what's happening is that the refresh token has expired.  There are two different tokens involved, the ID token and the refresh token.  If the ID token had expired, it would refresh it.  If the refresh token (long-lived) had expired, it will throw that error.

Here's some more about the refresh token and situations where it would expire:
https://firebase.google.com/docs/auth/admin/manage-sessions (Note this is from the Admin SDK docs.)

**Comment by cdock1029** [comment]
I'm am seeing this as well, but with the session cookie auth flow.

I get TOKEN_EXPIRED error trying to sign in ***if*** I just signed out within at least 30 or so seconds maybe more

### On server sign out

I'm using recommended revocation of refresh tokens when use signs out:

```typescript

// within server api route: /signout

...
const decodedClaims = await admin.auth().verifySessionCookie(sessionCookie)
await admin.auth().revokeRefreshTokens(decodedClaims.uid)
```

### client sign in
```typescript
async function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault()
    if (!busy) {
      setBusy(true)
      const target = e.target

      const email = target['email'].value
      const password = target['password'].value

      try {
        const cred = await auth.signInWithEmailAndPassword(email, password)
        const token = await cred.user?.getIdToken(true) /*  <-- fails  */
        console.log('TOKEN', token)

        await fetch('/api/polka/auth/login', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ token }),
        })
        try {
          await auth.signOut()
        } catch (e) {}
        setCookie('wpmauth', '1', 6)
        Router.replace('/')
      } catch (e) {
        setError(e.message)
      } finally {
        setBusy(false)
      }
    }
  }
```
I get a 400 error with this request:
`https://www.googleapis.com/identitytoolkit/v3/relyingparty/getAccountInfo?...`

and the call to `getIdToken` throws `TOKEN_EXPIRED`

Is there an expected delay before which a use cannot sign in after signing out with this flow?
I would expect `signInWithEmailAndPassword` to immediately create a new refresh token, and since this is literally within the process of signing in, the error instructing to sign in again isn't helpful, and a page refresh doesn't help.

**Comment by avolkovi** [comment]
@cdock1029 is it possible there is a race condition between when the user attempts to sign in and when the signOut is happening? Are you running your backend signOut asynchronously?

**Comment by cdock1029** [comment]
The /signout route on server waits for `revokeRefreshTokens` to complete before returning.

I think the reason the client-side error could even occur is because the revoke was successful.

```
const cred = await auth.signInWithEmailAndPassword(email, password)
const token = await cred.user?.getIdToken(true) /*  <-- fails  */
```
As noted above, the failure is client side when retrieving a token, but that's weird because the line right above it is a new sign in request.

And eventually this works. Whatever else is happening, I would expect an immediately new `signInwith..` request to produce a new refresh token.

**Comment by avolkovi** [comment]
@cdock1029 Sorry, maybe I was not clear.  Is it possible that the backend server is processing signOut at the same time you are trying to signIn?

signIn (in client)
... signOut (in server)
getIdToken (in client)

revoking refresh tokens works by setting the user's tokensValidAfterTime to the current time, if the signIn operation is happening concurrently with the token revocation it is possible for the token to be issued and then immediately be expired.  Can you please log the tokensValidAfterTime field in the backend after the revocation, and compare it to the value of the token's iat field (available on the token if calling getIdToken(false) to prevent forced refresh).

**Comment by cdock1029** [comment]
> possible that the backend server is processing signOut

From my point of view, only if the promise returned by `revokeRefreshTokens` resolved before the underlying operation to revoke the tokens completes.

The front end makes api call to /logout. When this returns (only after `revokeRefreshTokens` resolves), I redirect to login, then process login with form/client api.

For a short while, any repeated requests on client to signIn fail, even from incognito tab (possibly ruling out some sort of client side caching errors)

I have followed this https://firebase.google.com/docs/auth/admin/manage-cookies
and doing this
```ts
firebase.auth().setPersistence(firebase.auth.Auth.Persistence.NONE);
```
I have moved away from firebase auth for the moment though so this workflow has been removed. 
It could certainly have been a mistake somewhere in my code, no doubt. Might not be able to provide any more clarity at the moment.. simple reproduction would be to have server side logout that you are sure resolves before returning and see if login still works/fails.
